### ISSUE TITLE
关于Split Learning的使用，如果基于pytorch实现，对于SLModel的评估指标的计算逻辑

### ISSUE THREAD

**Issue by kylin5207** [secretflow/secretflow#1536]
### Issue Type

Others

### Source

binary

### Secretflow Version

secretflow 1.9.2b0

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.10

### What happend and What you expected to happen.
https://github.com/secretflow/secretflow/blob/main/tests/ml/nn/sl/test_sl_model_torch.py#L224
对于这里定义的base_model和fuse_model的metrics，具体的计算逻辑是怎样的。
```shell
base_model = TorchModel(
            model_fn=ConvNetBase,
            loss_fn=loss_fn,
            optim_fn=optim_fn,
            metrics=[
                metric_wrapper(
                    Accuracy, task="multiclass", num_classes=10, average='micro'
                ),
                metric_wrapper(
                    Precision, task="multiclass", num_classes=10, average='micro'
                ),
                metric_wrapper(AUROC, task="multiclass", num_classes=10),
            ],
        )

fuse_model = TorchModel(
            model_fn=ConvNetFuse,
            loss_fn=loss_fn,
            optim_fn=optim_fn,
            metrics=[
                metric_wrapper(
                    Accuracy, task="multiclass", num_classes=10, average='micro'
                ),
                metric_wrapper(
                    Precision, task="multiclass", num_classes=10, average='micro'
                ),
                metric_wrapper(AUROC, task="multiclass", num_classes=10),
            ],
        )
```
因为，这里模型的损失函数使用的是`nn.CrossEntropyLoss`，而且模型内部没有使用softmax将output转换为概率分布，所以模型的直接推理结果应该是未经softmax的数据。而Accuracy和Precision的计算又是针对于`argmax()`后的结果，这部分的逻辑是怎样的。

因为我在做二分类的任务，使用的损失函数为`nn.BCEWithLogitsLoss`，但是模型最终的AUROC计算结果很差（小于0.5），而损失函数呈下降趋势。

**Comment by kylin5207** [comment]
我的fuse model定义如下：

**Comment by kylin5207** [comment]
fuse_model = TorchModel(
    model_fn=FuseModel,
    loss_fn=loss_fn,
    optim_fn=optim_fn,
    metrics=[
        metric_wrapper(AUROC, task="binary"),
        metric_wrapper(Accuracy, task="binary"),
    ],
    input_dim=64,
    party_nums=2,
    output_dim=1,
    hidden_dim=64
)

**Comment by kylin5207** [comment]
loss的情况还可以，但是auc很差（小于0.5），用图像绘制
![auc](https://github.com/user-attachments/assets/c426083b-87c1-4d92-b0b9-2cd2b307238a)
![loss](https://github.com/user-attachments/assets/feef1339-0a80-458e-bdba-62621749c857)

**Comment by da-niao-dan** [comment]
hello 你好。
![image](https://github.com/user-attachments/assets/92e32b91-b542-41b4-9979-07359aa57918)
参考以上代码，pred_y只是 base model forward pass的结果。base model forward pass 没有 argmax 。
但是根据pytorch 文档，https://pytorch.org/ignite/generated/ignite.metrics.Accuracy.html
multiclass的时候，可以接受没有argmax的输入。

代码是正常的测试，不是调优的模型，你描述的现象是正常的。

**Comment by kylin5207** [comment]
好的，SLModel的`predict`是网络最后一层的直接输出咯。如果需要使用softmax处理后的预测结果，需要自己进行softmax计算。

**Comment by kylin5207** [comment]
我还有一个问题，虽然

> hello 你好。 ![image](https://private-user-images.githubusercontent.com/9532472/376164348-92e32b91-b542-41b4-9979-07359aa57918.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg5ODE5MDksIm5iZiI6MTcyODk4MTYwOSwicGF0aCI6Ii85NTMyNDcyLzM3NjE2NDM0OC05MmUzMmI5MS1iNTQyLTQxYjQtOTk3OS0wNzM1OWFhNTc5MTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMTVUMDg0MDA5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTNjZGU5MWZiZGVlNTUxOGZhM2QxYmZkMGI5MDg1YzQwOTcwMzAxZGNhMDQwNjNiYmE2Yzg5NmE4OWNhMDEzYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vNeb_P2wsucFtVAoIEhBUEBgW3Qa6EBS3xB_4L6u3T8) 参考以上代码，pred_y只是 base model forward pass的结果。base model forward pass 没有 argmax 。 但是根据pytorch 文档，https://pytorch.org/ignite/generated/ignite.metrics.Accuracy.html multiclass的时候，可以接受没有argmax的输入。
> 
> 代码是正常的测试，不是调优的模型，你描述的现象是正常的。

我还有一个问题，Accuracy的示例使用的模拟数据在[0,1]范围内，FuseModel的输出一定是[0,1]吗？

**Comment by kylin5207** [comment]
> hello， 没有argmax 不是说没有softmax， softmax 完了都是0，1 之间。如果模型定义了activation layer， 且是softmax， forward pass 会进行计算softmax的。

https://github.com/secretflow/secretflow/blob/main/tests/ml/nn/sl/model_def.py#L64看这里对于FuseModel的定义应该没有softmax呀

**Comment by da-niao-dan** [comment]
那的确Accuracy的输入在这里并不能保证是在0，1 时间。

**Comment by kylin5207** [comment]
> 那的确Accuracy的输入在这里并不能保证是在0，1 时间。

所以针对这个示例demo，就需要我上面说的，自己对predict的结果进行softmax计算

**Comment by da-niao-dan** [comment]
是的

**Comment by kylin5207** [comment]
好的感谢

**Comment by kylin5207** [comment]
> 是的

所以能基于tensorflow的二分类示例，给出pytorch实现的版本吗？https://github.com/secretflow/secretflow/blob/v1.9.0b2/docs/tutorial/Split_Learning_for_bank_marketing.ipynb

**Comment by kylin5207** [comment]
> 是的

为什么还没有问完，问题就被close了。
